caffe2/core/dispatch/DispatchTable.h

```diff
@@ -0,0 +1,150 @@
+    }
+
+private:
+    ska::flat_hash_map<Key, void*> map_;
+    // TODO Figure out how to get fast locking in C++11 (use boost::shared_timed_mutex? folly::SharedMutex?)
+    //mutable std::shared_timed_mutex mutex_;
+};
+} // namespace details
+
+/**
+ * Per-operator dispatch table.
+ *
+ * Given an operator specified by 'OpSchemaDef', this class records a dispatch table for
+ * various backends provided for this operator.  For example, if we consider the operator
+ * add(Tensor, Tensor), the dispatch table for this operator may contain implementations
+ * for various dynamic tensor types, such as (CPUFloatTensor, CPUFloatTensor),
+ * (CUDAFloatTensor, CUDAFloatTensor), etc.
+ *
+ * @tparam OpSchemaDef The operator signature this dispatch table encodes.
+ */
+// TODO: Support dispatch for meta-operators (which apply to all dynamic types)
+template<class OpSchemaDef>
+class DispatchTable final {
+private:
+  using Schema = OpSchema<OpSchemaDef>;
+
+public:
+  DispatchTable(): ops_() {}
+
+  /**
+   * Register an operator implementation in the table at some dispatch key.
+   * @param func Concrete function implementation to register
+   * @param dispatch_key Dispatch key to implement this function with
+   */
+  void registerOp(typename Schema::signature::func_type* func, typename Schema::dispatch::dispatch_key_type dispatch_key) {
+    ops_.emplace(std::move(dispatch_key), reinterpret_cast<void*>(func));
+  }
+
+  /**
+   * Unregister the operator implementation at some dispatch key.
+   *
+   * @param dispatch_key Dispatch key to unregister.
+   */
+  // TODO: This isn't going to work so well when we get more complicated override patterns!
+  // In this case, an operator will show up in multiple slots, and erasing them one-by-one
+  // is probably not such a good idea.
+  void deregisterOp(const typename Schema::dispatch::dispatch_key_type& dispatch_key) {
+    ops_.erase(dispatch_key);
+  }
+
+  /**
+   * Perform a dynamic dispatch on this table.
+   *
+   * @tparam Args Perfect forwarding template arguments to the dispatch
+   * @param args Arguments to invoke the function with
+   * @return Returned value of the operator
+   */
+  template<class... Args>
+  typename Schema::signature::return_type call(Args&&... args) const {
+    // TODO Better error message, but need to take care that reference arguments match non-reference arguments and so on.
+    //      static_assert(std::is_same<typename Schema::return_type (Args...), typename Schema::func_type>::value, "Argument types don't match operator signature");
+    auto operator_func = lookupOp_(args...);
+    return operator_func(std::forward<Args>(args)...);
+  }
+
+private:
+  template<class... Args>
+  typename Schema::signature::func_type* lookupOp_(const Args&... args) const {
+    auto dispatch_key = Schema::dispatch::dispatch_key(args...);
+    void* found = ops_.lookup(dispatch_key);
+    if (found == nullptr) {
+      throw std::logic_error("Didn't find operator to dispatch to");
```

Comment:
we need to include op name  and list of types

Reply:
We're not defining an op name in the schema (yet). I added a TODO to improve the error message here.

Reply:
At least the name feels like a must have functionality - why wait?

Reply:
ok, added it.